pkg/utils/logutil/log.go

```diff
@@ -88,22 +94,120 @@ func LogPanic() {
 	}
 }
 
+// RedactInfoLogType is the behavior of redacting sensitive information in logs.
+type RedactInfoLogType int
+
+const (
+	// RedactInfoLogOFF means log redaction is disabled.
+	RedactInfoLogOFF RedactInfoLogType = iota
+	// RedactInfoLogON means log redaction is enabled, and will replace the sensitive information with "?".
+	RedactInfoLogON
+	// RedactInfoLogMark means log redaction is enabled, and will use ‹› to enclose the sensitive information.
+	RedactInfoLogMark
+)
+
+// MarshalJSON implements the `json.Marshaler` interface to ensure the compatibility.
+func (t RedactInfoLogType) MarshalJSON() ([]byte, error) {
+	switch t {
+	case RedactInfoLogON:
+		return json.Marshal(true)
+	case RedactInfoLogMark:
+		return json.Marshal("MARK")
+	default:
+	}
+	return json.Marshal(false)
+}
+
+const invalidRedactInfoLogTypeErrMsg = `the "redact-info-log" value is invalid; it should be either false, true, or "MARK"`
+
+// UnmarshalJSON implements the `json.Marshaler` interface	to ensure the compatibility.
+func (t *RedactInfoLogType) UnmarshalJSON(data []byte) error {
+	var s string
+	if err := json.Unmarshal(data, &s); err == nil {
+		switch strings.ToUpper(s) {
+		case "FALSE", "OFF", "":
+			*t = RedactInfoLogOFF
+		case "MARK":
+			*t = RedactInfoLogMark
+		default:
+			*t = RedactInfoLogON
+		}
+		return nil
+	}
+	var b bool
+	if err := json.Unmarshal(data, &b); err == nil {
+		if b {
+			*t = RedactInfoLogON
+		} else {
+			*t = RedactInfoLogOFF
+		}
+		return nil
+	}
+	return errors.New(invalidRedactInfoLogTypeErrMsg)
+}
+
+// UnmarshalTOML implements the `toml.Unmarshaler` interface to ensure the compatibility.
+func (t *RedactInfoLogType) UnmarshalTOML(data any) error {
+	switch v := data.(type) {
+	case bool:
+		if v {
+			*t = RedactInfoLogON
+		} else {
+			*t = RedactInfoLogOFF
+		}
+		return nil
+	case string:
+		switch strings.ToUpper(v) {
+		case "FALSE", "OFF", "":
+			*t = RedactInfoLogOFF
+		case "MARK":
+			*t = RedactInfoLogMark
+		default:
+			*t = RedactInfoLogON
+		}
+		return nil
+	default:
+		return errors.New(invalidRedactInfoLogTypeErrMsg)
+	}
+}
+
 var (
-	enabledRedactLog atomic.Value
+	curRedactType atomic.Value
 )
 
 func init() {
-	SetRedactLog(false)
+	setRedactType(RedactInfoLogOFF)
+}
+
+func getRedactType() RedactInfoLogType {
+	return curRedactType.Load().(RedactInfoLogType)
 }
 
-// IsRedactLogEnabled indicates whether the log desensitization is enabled
-func IsRedactLogEnabled() bool {
-	return enabledRedactLog.Load().(bool)
+func setRedactType(redactInfoLogType RedactInfoLogType) {
+	curRedactType.Store(redactInfoLogType)
 }
 
-// SetRedactLog sets enabledRedactLog
-func SetRedactLog(enabled bool) {
-	enabledRedactLog.Store(enabled)
+const (
+	leftMark  = '‹'
+	rightMark = '›'
+)
+
+func redactInfo(input string) string {
```

Comment:
`redactInfo` just wraps the input with `< >`, it doesn't allow configuring specific strings to replace values in the logs. This is different from the logic in your code two weeks ago, is this what you expected?

Reply:
Yes, after discussion with PM, this is the only behavior when `MARK` is enabled.

Reply:
After checking the [tidb_redact_log](https://docs.pingcap.com/tidb/dev/system-variables#tidb_redact_log), it really is like that. I find it a bit strange, does it have a desensitizing effect?

Reply:
The purpose of this feature is to allow the cloud or customers to customize how to handle sensitive information. As for the output here in PD, sensitive information will not actually be desensitized; it is up to the user to decide how to handle this sensitive information.

Reply:
lgtm